### ISSUE TITLE
Review return types of matrix objects in linalg

### ISSUE THREAD

**Issue by rossbar** [networkx/networkx#4089]
The return types for various functions in linalg submodules are not consistent. For example, in the `laplacianmatrix` submodule, `laplacian_matrix` and `normalized_laplacian_matrix` return `scipy.sparse` matrices, whereas `directed_laplacian_matrix` and `directed_combinatorial_laplacian_matrix` return NumPy (dense) `matrix` objects.

I'm not sure if this is intentional or an implementation detail. If the latter, it's worth considering whether to make the return types of the various functions consistent with one another.

**Comment by dschult** [comment]
It seems like sparse matrices are the way to go for most network analysis, but I can also imagine users wanting to choose...
Still, it is easy enough to get a full array from a sparse representation.

I don’t recall any discussion of this implementation difference. I’m pretty sure it is not intentional.

**Comment by ericmjl** [comment]
Indeed, probably not.

That said, I think it’s probably better to provide sane defaults. My proposal is to default to returning sparse matrices, because memory usage, and allow some time to see if users instead want an kwarg in the API that allows returning of dense matrices instead:

```python
def adjacency_matrix_func(G, *args, **kwargs):
    # stuff
    mat = calculation_that_gives_sparse_matrix
    if return_dense_matrix=True:
        mat = sparse_matrix.to_dense()
    return mat
```

**Comment by dschult** [comment]
I agree that we should provide sane defaults...    And indeed sparse matrices seem like the best choice. 

I'd prefer to let the user switch to dense matrices using ```.to_dense()``` rather than crowd our function signature. but as you say-- we can see if users request a kwarg option.

**Comment by rossbar** [comment]
> I don’t recall any discussion of this implementation difference. I’m pretty sure it is not intentional.

Thanks for the feedback/historical perspective - I suspected this was the case but it's always hard to tell!

I also tend to lean towards always returning sparse matrices. This would also help deal with one hairy corner related to the `np.matrix` cleanup (see #3107): the `todense` method on sparse matrices returns an `np.matrix` object. In fact, this is probably one of the major reasons why the phase-out of `numpy.matrix` hasn't been more aggressive. I think it would definitely be preferable to avoid returning `np.matrix` objects from functions in networkx for the sake of consistency.

I am working on collating a list of functions that currently return `np.matrix` objects so that we can add `FutureWarnings` and change the return type, either to `ndarray`s or sparse matrices depending on what makes the most sense. I will add these functions to the list!

**Comment by ericmjl** [comment]
@rossbar to give this one a bit of momentum kick, once you have the list out, I think this can be one of those map-reduce-able newcomer tasks that we do on a bigger sprint. I think I can leverage this to put together some first-time contributor training material that we use in a live sprint.

**Comment by rossbar** [comment]
Great idea @ericmjl - I'll get that list put together either in #3107 and we can put a "good first issue" label or otherwise mark it for a future sprint.

**Comment by jarrodmillman** [comment]
I haven't had time to investigate this, so there may be nothing to do or you may have already looked into it.  This is mostly to make sure it is on your radar.

You should also take a close look at ``sparse/csgraph``:
- https://github.com/scipy/scipy/tree/master/scipy/sparse/csgraph

Ideally, we would use the same conventions.  We also don't use any of the functionality of ``sparse/csgraph`` and we should.  It may make sense to move some of the existing NX functionality/code there.  I don't have any specific suggestions (so this is just a meta-comment).

**Comment by willpeppo** [comment]
hey guys- as a new contributor who's just gone through a few first contributions / first sprint in the past month I'd be happy to help organize materials and/or an event.  I definitely have some thoughts after having just gone through this process.

**Comment by ericmjl** [comment]
@willpeppo, I just remembered, a bunch of us on the NetworkX team are going to be doing an "open office hours"-style thing where we can hack on this. Date is this Friday! 

If you're open to it, we should connect during the sprint hours! Let me know on LinkedIn if you can or cannot make it, and if you can, I'll let you know what the virtual meeting room link is that way.

**Comment by willpeppo** [comment]
awesome!  I would love to join.  Thanks, Eric.

**Comment by rossbar** [comment]
This issue was largely addressed by #5139 and related follow-up PRs. The question of whether functions return sparse or dense arrays really depends on whether a sparse or dense method is used under-the-hood. For instance, the internal `_transition_matrix` function is used in multiple places, and can return either a sparse or dense array depending on which `walktype` is used. By default, the `walktype` is selected based on the properties of the graph, which means the result can be sparse or dense depending on graph structure.

This can certainly be reviewed further, though I think most of the low-hanging fruit re: getting rid of `matrix` and updating to use array semantics has been covered.
